### ISSUE TITLE
tflite::Intepreter move constructor

### ISSUE THREAD

**Issue by catskul** [tensorflow/tensorflow#48908]
`tflite::Interpeter` doesn't have a move constructor.

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/interpreter.h

Because of this it's more difficult to make a resource pool of `Interpeter`s which is itself desirable because `Interpeter` is not itself threadsafe, and the critical piece which is needed once-per-thread, and is simultaneously expensive to build.

**System information**
- TensorFlow version (you are using): latest
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
Currently move is not possible. Move would is desirable.

**Will this change the current api? How?**
Adds constructor to `tflite::Interpeter` so technically it's an API change, but it's backward compatible.

**Who will benefit with this feature?**
Anyone attempting to use many `tflite::Interpeter`s in parallel.

**Any Other info.**
First pass glance at the code suggests that implementation likely be trivial.

**Comment by abattery** [comment]
@thaink could you take a look at this feature request?

**Comment by thaink** [comment]
@catskul Interpreter is almost always created as std::unique_ptr<Interpreter>. How about using that for the pool?

**Comment by catskul** [comment]
That certainly works, and is what I'm doing now, but the pool hands out handles which are effectively `unique_ptrs` and ends up creating an awkward double dereference, and causes generic code to need to special case `tflite::Intepreter` instead of treating it like all the other types.

Certainly not the end of the world not to have this, but it would be a nice-to-have.

**Comment by thaink** [comment]
@jdduke What do you think about this?

**Comment by catskul** [comment]
It makes sense, and is currently what I'm doing. It just forces an extra layer of indirection: `(*interpreterResource)->AllocateTensors()` where the code normally just does `interpreterResource->AllocateTensors()` isn't too bad on it's own, but where it intersects with templates/generic code which is expecting a single layer of indirection, then I need to special case with SFINAE which I'd rather not do when I don't have to.

NBD if this is seen as unnecessary churn, or is otherwise undesirable though. I'll can make do the way it is certainly.

With regard to how it would be used given the builder's interface requiring a `unique_ptr`. I was planning on moving the `Interpeter` out of the `unique ptr` this way:

```
// pseudo-code
auto creationFn = [&]() -> auto {
    auto myInterpeterPtr = unique_ptr<Interpeter>{};
    auto status = builder(&myInterpeterPtr);
    assert(status == kTfLiteOk);
    assert(myInterpeterPtr != nullptr);
    return std::move(*myInterpreterPtr);
}
auto resourcePool = ResourcePool(defaultNumOfResources, creationFn);
```

**Comment by catskul** [comment]
would a PR be welcome here? I'm don't currently have my dev env set up to make it easy to create/test appropriately to make a PR, but don't want to put in the effort if it's not wanted.

**Comment by jdduke** [comment]
> would a PR be welcome here? I'm don't currently have my dev env set up to make it easy to create/test appropriately to make a PR, but don't want to put in the effort if it's not wanted.

I'm sympathetic, but at this point, probably not. In part as I don't think we event want to have an Interpreter object left in an indeterminate/undefined state, and we also probably wouldn't want to expose a move assignment operator, so the use-case here is fairly narrow. If you hit issues with your particular application that you can't work around, we can revisit. At a technical level, I don't think the implementation of the move operator should be all that difficult. Thanks for the feature suggestion though!
